src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,270 @@
-  };
+  constructor() {
+    this.strings = [];
+    this.stringsMap = new Map<string, number>();
+    this.getIndexOrAdd('');
+  }
 
   /**
-   * Adds samples from a node and it's children to the fields tracking
-   * profile serialization.
-   *
-   * @param node - the node which is serialized
-   * @param stack - the stack trace to the current node.
+   * Returns index of str within the table. Also adds str to string table if
+   * str is not in the table already.
    */
-  function serializeNode(root: TimeProfileNode) {
-    let entries: Entry[] = [];
-    // don't include root node in serialized profile, start with it's children.
-    for (let child of root.children) {
-      entries.push({node: child, stack: []});
+  getIndexOrAdd(str: string): number {
+    let loc = this.stringsMap.get(str);
+    if (loc !== undefined) {
+      return loc;
     }
-    while (entries.length > 0) {
-      let entry = entries.pop();
-      if (entry !== undefined) {
-        let node = entry.node;
-        let stack = entry.stack;
-        let location = getLocation(node);
-        stack.unshift(location.id as number);
-        if (node.hitCount > 0) {
-          const sample = new perftools.profiles.Sample({
-            locationId: stack,
-            value: [node.hitCount, node.hitCount * intervalMicros]
-          });
-          samples.push(sample);
-        }
-        for (let child of node.children) {
-          entries.push({node: child, stack: stack.slice(0)});
-        }
+    loc = this.strings.push(str) - 1;
+    this.stringsMap.set(str, loc);
+    return loc;
+  }
+}
+
+/**
+ * Takes v8 profile and populates sample, location, and function fields of
+ * profile.proto.
+ *
+ * @param profile - profile.proto with empty sample, location, and function
+ * fields.
+ * @param root - root of v8 profile tree describing samples to be appended
+ * to profile.
+ * @param appendToSamples - function which converts entry to sample(s)  and
+ * appends these to end of an array of samples.
+ * @param stringTable - string table for the existing profile.
+ */
+function serialize(
+    profile: perftools.profiles.IProfile, root: ProfileNode,
+    appendToSamples: AppendEntryToSamples, stringTable: StringTable) {
+  const samples: Array<perftools.profiles.Sample> = [];
+  const locations: Array<perftools.profiles.Location> = [];
+  const functions: Array<perftools.profiles.Function> = [];
+  const locationMap: Map<number, perftools.profiles.Location> = new Map();
+  const functionMap: Map<number, perftools.profiles.Function> = new Map();
+  const functionIdMap = new Map<string, number>();
+  const locationIdMap = new Map<string, number>();
+
+  const entries: Entry[] = root.children.map((n) => ({node: n, stack: []}));
+  while (entries.length > 0) {
+    const entry = entries.pop();
+    if (entry !== undefined) {
+      const node = entry.node;
+      const stack = entry.stack;
+      const location = getLocation(node);
+      stack.unshift(location.id as number);
```

Comment:
Would "stack.unshift(location.id!);" work? As far as I understand that discards the "undefined | null" portion of the type which is more type-safe than a cast.

Reply:
No, I get the following error if I remove the cast:
[ts] Argument of type 'number | Long' is not assignable to parameter of type 'number'.
  Type 'Long' is not assignable to type 'number'